New Resource: aws_rds_reserved_instance

brittandeyoung opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #8521 

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

Replace ec2 with the service package corresponding to your tests.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make testacc TESTS=TestAccXXX PKG=ec2

...
```

brittandeyoung commented:
Currently trying to think through how to write acc tests for this type of resource as reservations cannot be deleted and are valid for the duration of the specified offering id once created.

YakDriver commented:
@brittandeyoung Thanks for your work on this! After a quick skim, it's looking very promising.

We've also been thinking about how to test. Testing is important but there is precedent for things we simply cannot test. Rather than hold something up indefinitely, we may have to do best effort and rely on the community for reports of problems.

In thinking about full support for reserved instances, I'm going to plunk down this list here for future reference:

* [AcceptReservedInstancesExchangeQuote](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.AcceptReservedInstancesExchangeQuote)
* [CancelReservedInstancesListing](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.CancelReservedInstancesListing)
* [CreateReservedInstancesListing](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.CreateReservedInstancesListing)
* [DeleteQueuedReservedInstances](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DeleteQueuedReservedInstances)
* [DescribeReservedInstances](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeReservedInstances)
* [DescribeReservedInstancesListings](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeReservedInstancesListings)
* [DescribeReservedInstancesModifications](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeReservedInstancesModifications)
* [DescribeReservedInstancesOfferings](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeReservedInstancesOfferings)
* [GetReservedInstancesExchangeQuote](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.GetReservedInstancesExchangeQuote)
* [ModifyReservedInstances](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.ModifyReservedInstances)
* [PurchaseReservedInstancesOffering](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.PurchaseReservedInstancesOffering)

bschaatsbergen commented on website/docs/r/rds_reserved_instance.html.markdown:
```diff
@@ -0,0 +1,60 @@
+* `start_time` - The time the reservation started.
+* `state` - The state of the reserved DB instance.
+* `usage_price` - The hourly price charged for this reserved DB instance.
+* `tags_all` - A map of tags assigned to the resource, including those inherited from the provider [`default_tags` configuration block](/docs/providers/aws/index.html#default_tags-configuration-block).
```
Drive-by suggestion

```suggestion
* `tags_all` - A map of tags assigned to the resource, including those inherited from the provider [`default_tags` configuration block](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#default_tags-configuration-block).
```

brittandeyoung replied:
Suggestion has been implemented.

bschaatsbergen commented on website/docs/r/rds_reserved_instance.html.markdown:
````diff
@@ -0,0 +1,60 @@
+## Example Usage
+
+```terraform
+resource "aws_rds_reserved_instance" "my-reservation" {
````
Drive-by suggestion

```suggestion
resource "aws_rds_reserved_instance" "example" {
```

brittandeyoung replied:
Suggestion has been implemented.

bschaatsbergen reviewed (commented):
Nice work :+1: :eyes:

YakDriver commented on internal/service/rds/consts.go:
```diff
@@ -155,3 +155,7 @@ func TimeoutAction_Values() []string {
 )
+
+const (
+	ResTags = "Tags"
```
```suggestion
	ResNameTags = "Tags"
```

brittandeyoung replied:
Suggestion has been implemented.

YakDriver commented on internal/service/rds/reserved_instance.go:
```diff
@@ -0,0 +1,215 @@
+)
+
+const (
+	ResReservedInstance = "Reserved Instance"
```
```suggestion
	ResNameReservedInstance = "Reserved Instance"
```

brittandeyoung replied:
Suggestion has been implemented.

brittandeyoung commented:
There has been some discussion with the provider team about this potentially needing to be written as a framework resource instead of using the standard SDK method. I am currently waiting on direction from the team on this, and will make required modifications when I hear back.

zhelanov commented:
Hi @brittandeyoung , any update regarding the feature? It's awesome and I think that our team is not the only one who's waiting for it

brittandeyoung commented:
@zhelanov Thank you for the bump on this PR. I have been working on a few other contributions as well and forgot about this one. I will be this updated to resolve the conflicts and have passing tests. Then wait for the provider team to review.

brittandeyoung commented:
@YakDriver Sorry for the delay in getting back to this PR. It has been updated with all recommendations and checks are all passing.

YakDriver commented on website/docs/r/rds_reserved_instance.html.markdown:
````diff
@@ -0,0 +1,60 @@
+## Example Usage
+
+```terraform
+resource "aws_rds_reserved_instance" "example" {
+  offering_id    = "438012d3-4052-4cc7-b2e3-8d3372e0e706"
+  reservation_id = "optionalCustomReservationID"
+  instance_count = 3
+}
````
```suggestion
data "aws_rds_reserved_offering" "example" {
  db_instance_class = "db.r5.large"
}

resource "aws_rds_reserved_instance" "example" {
  offering_id    = data.aws_rds_reserved_offering.example.offering_id
  reservation_id = "optionalCustomReservationID"
  instance_count = 3
}
```

For tests, something like this should work.

YakDriver commented:
Here's a regex that can be helpful for finding db instance classes in the tests and docs:

```
db\.[a-z]{1,2}\d[a-z]?\.[a-z0-9]{0,4}(large|medium|small|micro|nano)
```

brittandeyoung commented:
I have added a data source for `aws_rds_reserved_instance_offering` to the pull request with passing tests. Working on making the tests for the reservation now. 

The cheapest test we will be able to run is the 102$ up front instance offering.

```
make testacc TESTARGS='-run=TestAccRDSInstanceOffering_' PKG_NAME=internal/service/rds  
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/rds/... -v -count 1 -parallel 20  -run=TestAccRDSInstanceOffering_ -timeout 180m
=== RUN   TestAccRDSInstanceOffering_basic
=== PAUSE TestAccRDSInstanceOffering_basic
=== CONT  TestAccRDSInstanceOffering_basic
--- PASS: TestAccRDSInstanceOffering_basic (11.80s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/rds        13.670s
```

brittandeyoung commented:
@YakDriver I have added the `aws_rds_reserved_instance_offering` datasource as recommended with passing tests, and added tests for the `aws_rds_reserved_instance` resource that uses the new datasource and will only run when the enviornment variable `RUN_RDS_RESERVED_INSTANCE_TESTS` is set to `true`. 

This test will cost around $102 per run and vary based on the region testing is done in. I do not have an AWS account where I can simply rack up a $102 charge each time I attempt to validate my tests =) . Does the Hashi team have an account where these tests can be validated in? 

At this point my code is ready for another review.

brittandeyoung commented:
Something is up with the `Pull Request Target (All types) / PullRequestComments (pull_request_target) ` check. False failures

YakDriver reviewed (approved):
Looks great! 🎉
